p2p/discovery/discovery.go

```diff
@@ -41,17 +61,39 @@ type Discovery struct {
 	disc Protocol
 
 	local        *node.LocalNode
-	rt           *addrBook
-	bootstrapper *refresher
+	rt           addressBook
+	bootstrapper bootstrapper
 }
 
 // Size returns the size of addrBook.
 func (d *Discovery) Size() int {
 	return d.rt.NumAddresses()
 }
 
+func (d *Discovery) Good(key p2pcrypto.PublicKey) {
+	d.rt.Good(key)
+}
+
+func (d *Discovery) Attempt(key p2pcrypto.PublicKey) {
+	d.rt.Attempt(key)
+}
+
+func (d *Discovery) refresh(ctx context.Context, peersToGet int) error {
+	err := d.bootstrapper.Bootstrap(ctx, peersToGet)
+	if err != nil {
+		d.local.Log.Error("Error trying to refresh addrbook ", err)
+		return err
+	}
+	return nil
+}
+
 // SelectPeers asks routing table to randomly select a slice of nodes in size `qty`
 func (d *Discovery) SelectPeers(qty int) []*node.NodeInfo {
+
+	if d.rt.NeedNewAddresses() {
+		d.refresh(context.Background(), qty) // TODO: use ctx with timeout, check errors
```

Comment:
The options I though on we're:
- call it periodically in the background. means we'll do it all the time.
- call it when a peer removed/changed state(bad/tried), requires to track the size and properties of the routing table in each action
- call it in the background when asking for peers and we have missing - we'll not get the new peers as a response and we'll have to remember to call again when refresh is done.
I thought this way is simpler. we're not blocking anything other than selecting peers, and we mandatorily need more peers so it is ok.

Reply:
Not sure I understand what is wrong with the first option. Why not calling it every period of time? Most of the times it will check if we are missing peers, see that we have enough and go back to sleep. Only on rare ocations, it will actually fetch new peers.

Reply:
Okay, so we should have a side note about that we chose that strategy for now.